### DISCUSSION TITLE
Using react-number-format in v7, cannot read property value?

### DISCUSSION THREAD

**Discussion by c0d3x** [react-hook-form/react-hook-form#5194] [Q&A] [answered]
Trying to use CurrencyInput component in my form, earlier I did not have to use the Controller component with ref={register} in v6.

Now I cant seem to get it to work after bumping to v7, tried using controller to same issue with value.

![image](https://user-images.githubusercontent.com/661911/118271752-fd856e00-b4c1-11eb-9a2c-e9a8974064dd.png)

Here is an example, any clues?
https://codesandbox.io/s/react-hook-form-numberformat-gez66?file=/src/index.tsx

**Comment by lamhungypl** [comment]
I leave `onChange` does its job and manual `setValue` to form (both v6 and v7 worked!).


```javascript
<Controller
  render={({ field: { onChange, ...renderProps } }) => {
    return (
      <NumberFormat
        {...renderProps}
        onValueChange={target => {
          onChange();
          setValue('price', target.floatValue);
        }}
        isNumericString
        decimalScale={2}
        fixedDecimalScale
        allowNegative={false}
        prefix="$"
      />
    );
  }}
  name='price'
  defaultValue={record.price}
  control={control}
  rules={{
    
  }}
/>
```

**Reply by c0d3x** [reply]
Is it possible to get the value without the controlled Controller component?
I dont want react-hook-form deps in my CurrencyComponent, as this will be in its own npm registry and I dont want to maintain react-hook-form versions in multiple places.. Also I dont want to wrap every CurrencyInput with a Controller..

**Reply by lamhungypl** [reply]
@c0d3x  I think `useController` is what you need for reuse Currency component, then `errorText` or things related `form`, check the `useFormContext`. 

https://codesandbox.io/s/react-hook-form-numberformat-forked-kyut2?file=/src/CurrencyInputClean.tsx  
```javascript
export const CurrencyInputClean = ({ name, control }) => {
  const { field } = useController({ name, control });
  const { onChange, ...rest } = field;
  const [currency, setCurrency] = useState();

  return (
    <div>
      <NumberFormat
        {...rest}
        className="form-control"
        allowNegative={false}
        isNumericString
        prefix="$ "
        suffix=",-"
        thousandSeparator=" "
        decimalSeparator="."
        decimalScale={0}
        onValueChange={(value) => {
          onChange(value.floatValue);
        }}
      />
    </div>
  );
};

```

**Comment by leapful** [comment, marked as answer]
@c0d3x 

You are using wrong params for "onChange" of react-hook-form in CurrencyInputClean component, the "onChange" expects a native event which can be simulated with same object structure like:

```
onChange({
  type: 'text',
  target: {
     name, // form element "name" attribute
     value: v.floatValue
  }
})
``` 

Working sample: https://codesandbox.io/s/react-hook-form-numberformat-forked-w63c2?file=/src/CurrencyInputClean.tsx

**Reply by c0d3x** [reply]
Yes, this!

**Reply by bluebill1049** [reply]
thank you @leapful & @lamhungypl ❤️  for the help.

**Reply by c0d3x** [reply]
![image](https://user-images.githubusercontent.com/661911/118643287-a1318f80-b7dc-11eb-99af-d1f6e4be2b2a.png)
However the defaultValue is not getting trough to the component, any clue on that?
https://codesandbox.io/s/react-hook-form-numberformat-forked-lhmvw?file=/src/index.tsx

**Reply by leapful** [reply]
@c0d3x 

You are missing "defaultValue" when rendering CurrencyInputClean component in index.tsx.

Working sample: https://codesandbox.io/s/react-hook-form-numberformat-forked-khjio?file=/src/CurrencyInputClean.tsx

**Reply by c0d3x** [reply]
Thanks for your reply!
Seems that it works setting the defaultValue on each field, so at least it is a working fix.

But this does not seem to be the best solution, having to use watch and define all the the values and then adding them to each field manually to defaultValue.

I have always added them to defaultValues in useForm and that has worked fine until i bumped to v7.
Do you know why the component is not getting the defaultValue from useForm defaultValues?

**Reply by leapful** [reply]
@c0d3x 

By default, "react-hook-form" sets the input default value correctly but it was reset because of re-rendering on "react-number-format" NumberFormat. This Class component is calling "setState" inside componentDidMount() life cycle hook, therefore it will always trigger input element DOM updating on every first reconciliation of React but at that time, the whole props data (including empty "value") is old data from the previous rendering, therefore, it will override the updated default value set by "react-hook-form".

"react-hook-form" v6 has different ref handling, that's why it don't have this issue.

Referrer: https://github.com/s-yadav/react-number-format/blob/master/src/number_format.js#L126-L132
